ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,185 @@
+// The taskManifestResponder will ack the task manifest message and send ACS a
+// TaskStopVerificationMessage if there are stop candidates.
+type taskManifestResponder struct {
+	taskComparer      TaskComparer
+	snAccessor        SequenceNumberAccessor
+	messageIDAccessor ManifestMessageIDAccessor
+	respond           wsclient.RespondFunc
+	metricsFactory    metrics.EntryFactory
+}
+
+// NewTaskManifestResponder returns an instance of the taskManifestResponder struct.
+func NewTaskManifestResponder(taskComparer TaskComparer, snAccessor SequenceNumberAccessor, messageIDAccessor ManifestMessageIDAccessor,
+	metricsFactory metrics.EntryFactory, responseSender wsclient.RespondFunc) *taskManifestResponder {
+	r := &taskManifestResponder{
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	logger.Debug(fmt.Sprintf("Processing %s with MessageId %s", TaskManifestMessageName, aws.StringValue(message.MessageId)))
+	m := tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration)
+	err := tmr.processTaskManifestMessage(message)
+	m.Done(err)
+}
+
+// processTaskManifestMessage processes a task manifest message from ACS. It verifies
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower or equal to to the latest known one, the TaskManifestMessage must be discarded.
+	// The manifest should also not be a duplicate of the last one received, so the number cannot be less than or equal to the latest.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum <= agentLatestSeqNum {
+		logger.Warn(fmt.Sprintf("Discarding task manifest message with id %s "+
+			"Sequence number from manifest is less than or equal to Agent latest sequence number.", aws.StringValue(message.MessageId)), logger.Fields{
```

Comment:
Could we put the message ID in the `logger.Fields` along with `field.MessageID` and in this log message?

Reply:
ack